Projects/Server/Common/Filters/Filters.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+
+namespace Server.Common.Filters;
+
+public readonly struct DummyFilter<T> : IFilter<T, T>
+{
+    public T? Invoke(T instance)
+    {
+        return instance;
+    }
+}
+
+public readonly struct ChainFilter<TIn, TMiddle, TOut, TFirst, TSecond> : IFilter<TIn, TOut>
+        where TFirst : IFilter<TIn, TMiddle>
+        where TSecond : IFilter<TMiddle, TOut>
+{
+    private readonly TFirst first;
+    private readonly TSecond second;
+
+    public ChainFilter(TFirst first, TSecond second)
+    {
+        this.first = first;
+        this.second = second;
+    }
+
+    public TOut? Invoke(TIn instance)
+    {
+        return first.Invoke(instance) is not { } middle ? default : second.Invoke(middle);
+    }
+}
+
+public readonly struct PredicateFilter<T> : IFilter<T, T>
+{
+    private readonly Predicate<T> predicate;
```

Comment:
We might want to do specific benchmarking on this. I wonder about JIT optimization here since I think even using a static `Predicate<T>` might be really slow to execute. If we don't have lots of use cases, maybe someone should create a specific predicate-style filter for this like we do with Comparers.